### PR TITLE
[ refactor ] contexts as snoclists of types

### PR DESCRIPTION
**Pull request by gallais**

### CODE REVIEW

File: tests/working/003-shadowing/expected [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Running Test
+# Finished Parsing
+# Finished Type Checking
+a : Nat
+a : Bool
+a : Bool
+a : Nat
+a : Nat -> Nat
```
Comment (gallais) [inline tests/working/003-shadowing/expected]: Should the printing of holes drop shadowed names?

Comment (jfdm) [reply tests/working/003-shadowing/expected]: probably...

**Comment by gallais** [comment]
Can I merge this? CSE will build on top of it.

**Comment by jfdm** [comment]
Of course! I was ignoring this as I assumed that you would be merging it when *you* thought it was appropriate to merge.

**Pull request merged**